File: apps/clinical-engine/src/ehas2_clinical_engine/rule4/severity/severity_fingerprint_v1.py
from __future__ import annotations

import hashlib

from ..canonical_json import canonical_stable_dumps

RULE4_SEVERITY_RESOLUTION_FINGERPRINT_V1 = "rule4-severity-resolution-fingerprint-v1"


def _slot_payload(slot: dict) -> dict:
    return {
        "formula_slot_id": slot["formula_slot_id"],
        "formula_target_id": slot["formula_target_id"],
        "target_role": slot["target_role"],
        "severity_status": slot["severity_status"],
        "severity_score": slot["severity_score"],
        "severity_band": slot["severity_band"],
        "severity_resolution_source": slot["severity_resolution_source"],
        "binding_status": slot["binding_status"],
        "evidence_item_ids": sorted(slot["evidence_item_ids"]),
        "corroborating_source_ids": sorted(slot["corroborating_source_ids"]),
        "selected_cascade": slot["selected_cascade"],
        "selected_dilution": slot["selected_dilution"],
        "upstream_context_status": slot["upstream_context_status"],
        "reason_codes": sorted(slot["reason_codes"]),
        "limitation_codes": sorted(slot["limitation_codes"]),
    }


def rule4_severity_resolution_fingerprint_v1_payload(
    *,
    ruleset_version: str,
    registry_version: str,
    slot_resolutions: list[dict],
    reason_codes: list[str],
    limitation_codes: list[str],
) -> str:
    slots = sorted(slot_resolutions, key=lambda s: s["formula_slot_id"])
    return canonical_stable_dumps(
        {
            "fingerprint_version": RULE4_SEVERITY_RESOLUTION_FINGERPRINT_V1,
            "ruleset_version": ruleset_version,
            "registry_version": registry_version,
            "slot_resolutions": [_slot_payload(s) for s in slots],
            "reason_codes": sorted(set(reason_codes)),
            "limitation_codes": sorted(set(limitation_codes)),
        }
    )


def rule4_severity_resolution_fingerprint_v1_hash(
    *,
    ruleset_version: str,
    registry_version: str,
    slot_resolutions: list[dict],
    reason_codes: list[str],
    limitation_codes: list[str],
) -> str:
    payload = rule4_severity_resolution_fingerprint_v1_payload(
        ruleset_version=ruleset_version,
        registry_version=registry_version,
        slot_resolutions=slot_resolutions,
        reason_codes=reason_codes,
        limitation_codes=limitation_codes,
    )
    return hashlib.sha256(payload.encode("utf-8")).hexdigest().upper()


def fingerprint_from_severity_output(output: dict) -> str:
    return rule4_severity_resolution_fingerprint_v1_hash(
        ruleset_version=output["ruleset_version"],
        registry_version=output["registry_version"],
        slot_resolutions=list(output["slot_resolutions"]),
        reason_codes=list(output["reason_codes"]),
        limitation_codes=list(output["limitation_codes"]),
    )

File: packages/clinical-contracts/src/rule4/severity/severityFingerprintV1.ts
import { createHash } from 'node:crypto';

import { canonicalStableDumps } from '../canonicalJson.js';
import type { Rule4SeverityAdapterOutput, Rule4SlotSeverityResolution } from './types.js';

export const RULE4_SEVERITY_RESOLUTION_FINGERPRINT_V1 =
  'rule4-severity-resolution-fingerprint-v1' as const;

function slotPayload(slot: Rule4SlotSeverityResolution): Record<string, unknown> {
  return {
    formula_slot_id: slot.formulaSlotId,
    formula_target_id: slot.formulaTargetId,
    target_role: slot.targetRole,
    severity_status: slot.severityStatus,
    severity_score: slot.severityScore,
    severity_band: slot.severityBand,
    severity_resolution_source: slot.severityResolutionSource,
    binding_status: slot.bindingStatus,
    evidence_item_ids: [...slot.evidenceItemIds].sort(),
    corroborating_source_ids: [...slot.corroboratingSourceIds].sort(),
    selected_cascade: slot.selectedCascade,
    selected_dilution: slot.selectedDilution,
    upstream_context_status: slot.upstreamContextStatus,
    reason_codes: [...slot.reasonCodes].sort(),
    limitation_codes: [...slot.limitationCodes].sort(),
  };
}

export function rule4SeverityResolutionFingerprintV1Payload(input: {
  rulesetVersion: string;
  registryVersion: string;
  slotResolutions: readonly Rule4SlotSeverityResolution[];
  reasonCodes: readonly string[];
  limitationCodes: readonly string[];
}): string {
  const slots = [...input.slotResolutions]
    .sort((a, b) => a.formulaSlotId.localeCompare(b.formulaSlotId))
    .map(slotPayload);
  return canonicalStableDumps({
    fingerprint_version: RULE4_SEVERITY_RESOLUTION_FINGERPRINT_V1,
    ruleset_version: input.rulesetVersion,
    registry_version: input.registryVersion,
    slot_resolutions: slots,
    reason_codes: [...new Set(input.reasonCodes)].sort(),
    limitation_codes: [...new Set(input.limitationCodes)].sort(),
  });
}

export function rule4SeverityResolutionFingerprintV1Hash(input: {
  rulesetVersion: string;
  registryVersion: string;
  slotResolutions: readonly Rule4SlotSeverityResolution[];
  reasonCodes: readonly string[];
  limitationCodes: readonly string[];
}): string {
  const payload = rule4SeverityResolutionFingerprintV1Payload(input);
  return createHash('sha256').update(payload, 'utf8').digest('hex').toUpperCase();
}

export function fingerprintFromSeverityOutput(
  output: Pick<
    Rule4SeverityAdapterOutput,
    'rulesetVersion' | 'registryVersion' | 'slotResolutions' | 'reasonCodes' | 'limitationCodes'
  >,
): string {
  return rule4SeverityResolutionFingerprintV1Hash({
    rulesetVersion: output.rulesetVersion,
    registryVersion: output.registryVersion,
    slotResolutions: output.slotResolutions,
    reasonCodes: output.reasonCodes,
    limitationCodes: output.limitationCodes,
  });
}
